idx-kotlin/src/main/java/com/okta/idx/kotlin/dto/v1/RemediationMiddleware.kt

```diff
@@ -0,0 +1,142 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.okta.idx.kotlin.dto.v1
+
+import com.okta.idx.kotlin.dto.IdxAuthenticatorCollection
+import com.okta.idx.kotlin.dto.IdxMessageCollection
+import com.okta.idx.kotlin.dto.IdxRemediation
+import com.okta.idx.kotlin.dto.IdxRemediationCollection
+
+internal fun Response.toIdxRemediationCollection(): IdxRemediationCollection {
+    val remediations = mutableListOf<IdxRemediation>()
+
+    if (remediation?.value != null) {
+        for (form in remediation.value) {
+            remediations += form.toIdxRemediation()
+        }
+    }
+
+    if (cancel != null) {
+        remediations += cancel.toIdxRemediation()
+    }
+
+    if (successWithInteractionCode != null) {
+        remediations += successWithInteractionCode.toIdxRemediation()
+    }
+
+    return IdxRemediationCollection(remediations)
+}
+
+private fun Form.toIdxRemediation(): IdxRemediation {
+    val form = IdxRemediation.Form(value?.map { it.toIdxField() } ?: emptyList())
+    val remediationType = name.toRemediationType()
+    return IdxRemediation(
+        type = remediationType,
+        name = name,
+        form = form,
+        authenticators = IdxAuthenticatorCollection(emptyList()), // TODO:
+        method = method,
+        href = href,
+        accepts = accepts,
+        refresh = null, // TODO:
+        relatesTo = relatesTo
+    )
+}
+
+private fun FormValue.toIdxField(): IdxRemediation.Form.Field {
+    // Fields default to visible, except there are circumstances where
+    // fields (such as `id`) don't properly include a `visible: false`. As a result,
+    // we need to infer visibility from other values.
+    var actualVisibility = visible ?: true
+    if (mutable == false && value != null) {
+        actualVisibility = false
+    }
+
+    // TODO: Use middleware context?
+    // TODO: There are some odd interactions how form/value is handled. Might need to do some conversions here.
+
+    return IdxRemediation.Form.Field(
+        name = name,
+        label = label,
+        type = type,
+        _value = value,
+        isVisible = actualVisibility,
+        isMutable = mutable ?: true,
+        isRequired = required ?: false,
+        isSecret = secret ?: false,
+        relatesTo = relatesTo,
+        form = form?.value?.toForm(),
+        options = options?.map { it.toIdxField() },
+        messages = IdxMessageCollection(messages?.value?.map { it.toIdxMessage() } ?: emptyList()),
+        authenticator = null, //TODO: Might not be able to initialize here, might need to do a late init.
+    )
+}
+
+private fun String.toRemediationType(): IdxRemediation.Type {
+    return when (this) {
+        "identify" -> IdxRemediation.Type.IDENTIFY
```

Comment:
So whenever Okta backend introduces a new remediation type, we need to maintain this mapping here and elsewhere in this SDK?

Reply:
Yeah, the idea here is that the backend statuses don't map directly to the sdk types. Say we have a V2, we could maintain our mapping internally, without breaking users of the SDK.

That being said, if we don't have the mapping because it's new or otherwise, the user can access to the type directly via another field. This is mostly just a nice to have thing.